### ISSUE TITLE
4.042 breaks LONGBLOB contents with mysql_enable_utf8

### ISSUE THREAD

**Issue by mgruner** [perl5-dbi/DBD-mysql#107]
@mbeijen, @pali: While working on our solution for #106, I noted a different problem.

With `mysql_enable_utf8` enabled, if we send binary data (only octets, not unicode) into a LONGBLOB column and retrieve it back, it is not the same as what we stored. Could that be caused by DBD::mysql trying to decode it as UTF-8? Or would that be intended?

**Comment by pali** [comment]
Yes, when `mysql_enable_utf8` is enabled then all input data from Perl to DBD::mysql are treated as wide (unicode) strings. The only exception is bind parameters of binary types: `SQL_BIT`, `SQL_BLOB`, `SQL_BINARY`, `SQL_VARBINARY` and `SQL_LONGVARBINARY.

**Comment by pali** [comment]
As perl itself do not know if particular scalar contains wide (unicode) string, binary data or UTF-8 encoded string, caller is responsible to know it. And the only way how to tell such information to DBI is via `bind()`. Basically ascii string (where each byte has highest bit zero) is same for binary data, UTF-8 and Latin1. So even with wrong "type" everything works... But when there are characters above 0x7F it starts to broken as encoding such characters to Latin1 and UTF-8 produces different (binary) data.

**Comment by pali** [comment]
And due to limitation of MySQL protocol DBI does not know type of some column at time of binding supplied parameters...

**Comment by mgruner** [comment]
@pali, I don't understand that, sorry. We don't send wide strings, just octets. The problem is when we fetch data from the DB. Did I miss something?

**Comment by pali** [comment]
@mgruner There is no difference between scalar which contains wide strings and scalar which contains octets. Perl does not distinguish between them. Let's say we have `my $a = "\x10"` and `my $b = "\N{U+10}"`. You could say that `$a` maybe contains just octets and `$b` just wide string. But that is not truth, both scalars `$a` and `$b` are same and `$a cmp $b` is truth. In both cases in scalar is stored character with ordinal number 10. Perl do not see differences between them and (correctly written) XS code too.

So if you say "I stored binary octets into scalar" you need to tell this fact somehow also to function into which you passed that scalar. Which means DBD::mysql does not know if you are passing octets or wide strings. DBD::mysql or any perl code just know it is scalar which contains sequence of ordinal numbers... Due to this fact there is third parameter for DBI `bind` to specify if your scalar should be treated as binary octets or as (wide) string.

It is more clear now? It is not DBI specific, but it is how perl itself is working and interpreting strings and Unicode...

**Comment by mgruner** [comment]
@pali thanks for the explanation. I'm not sure if I can fully follow, though. If I use DBD::mysql without `mysql_enable_utf8`, it throws me errors when I pass in a wide string. Perl is definitely able to tell the difference because of the utf8 flag. I'm not sure about XS code, I have no experience with that.

**Comment by pali** [comment]
> Perl is definitely able to tell the difference because of the utf8 flag.

No it cannot. Reason is what `SVf_UTF8` flag means. It does *not* say if string is wide Unicode or binary octets! `utf8` is absolutely criplic name and lot of people are confused and think something different because of name.

> If I use DBD::mysql without `mysql_enable_utf8`, it throws me errors when I pass in a wide string.

Look at https://metacpan.org/pod/DBI#bind_param it has optional third argument for specifying type. If you need to pass binary octets with `mysql_enable_utf8` into DBD::mysql you need to specify binary type (see my first comment).

Or when `mysql_enable_utf8` is not enabled and you need to pass Unicode wide string, use: `my $octets = Encode::encode("UTF-8", $wide_string)` and then pass `$octets` into DBI (plus `SET NAMES utf8` is needed).

> I'm not sure if I can fully follow, though.

If is something still not clear, write what or which part.

> I'm not sure about XS code, I have no experience with that.

You do not need to understand XS. If it is correctly written (which DBD::mysql now claims to be) it acts exactly same as pure perl code.

**Comment by mgruner** [comment]
Hello @pali. Thanks a lot for your patience and explanations!

> When set, any statement or bind parameter which is not of binary type is automatically encoded to UTF-8 octets before being sent to the MySQL server.

I still don't understand how DBD::mysql should try to encode something that is already encoded as a sequence of octets, rather than a wide string. To do so, first a `Decode()` call would have to be made to turn it into a wide string. But from what charset? From UTF8? Then the whole decode-encode sequence could be skipped. And from another charset is not really possible as it would not be known.

I'll close this issue now as it seems to work as designed.

**Comment by pali** [comment]
> I still don't understand how DBD::mysql should try to encode something that is already encoded as a sequence of octets, rather than a wide string.

There is no difference for perl between scalar encoded as (wide) string and scalar encoded as sequence of octets. In both cases scalar contains sequence of ordinal numbers. Just wide one could (but does not have to!) contain ordinal numbers above 255. So DBD::mysql does not if something is already encoded or not.

So if you have wide string (sequence of number) and you want to send via byte protocol (over network) you need to encode them into consistent encoding which have all bytes (sequence of numbers) below 256. One of option is UTF-8 (8bite Unicode Transformation Format).

So you convert scalar which contains that wide string into new scalar which would contain only bytes (UTF-8). But for perl there is no flag or any other information to know which (old or new?) scalar contains bytes and which could potentially contain also wide characters. So you as caller needs to know which one is byte-orientated and which one character-orientated.

And so when passing it to DBD::mysql you need to tell DBI if scalar should be treated as characters or if it should be treated as bytes.

> To do so, first a `Decode()` call would have to be made to turn it into a wide string.

No. You need to tell DBD::mysql that scalar which I'm passing "right now" is character-orientated or byte-orientated.

Seems you still have not caught it. There must not be any decode-encode-decode-encode sequences... Maybe for understanding it would help you to write (on paper :-)) how some wide string looks like and how after encoding them to UTF-8. In both cases it is just sequence of numbers. Always you should look at those scalars as on sequence of (arbitrary) numbers.

Basically what scalar contains you can print by this subroutine:
`sub show_scalar { say join ",", map ord, split //, $_[0] }`

**Comment by pali** [comment]
See documentation update in https://github.com/perl5-dbi/DBD-mysql/pull/119 how to temporary workaround buggy `mysql_enable_utf8` for older DBD::mysql verions (< 4.042).
